google-cloud-storage/src/main/java/com/google/cloud/storage/BlobId.java

```diff
@@ -117,14 +119,13 @@ public static BlobId of(String bucket, String name, Long generation) {
    * @param gsUtilUri the Storage url to create the blob from
    */
   public static BlobId fromGsUtilUri(String gsUtilUri) {
-    if (!Pattern.matches("gs://.*/.*", gsUtilUri)) {
+    Matcher m = gsUtilUriPattern.matcher(gsUtilUri);
+    if (!m.matches()) {
       throw new IllegalArgumentException(
-          gsUtilUri + " is not a valid gsutil URI (i.e. \"gs://bucket/blob\")");
+          gsUtilUri + " is not a valid gsutil URI (i.e. \"gs://bucket/blob\" or \"gs://bucket/blob#generation\")");
     }
-    int blobNameStartIndex = gsUtilUri.indexOf('/', 5);
-    String bucketName = gsUtilUri.substring(5, blobNameStartIndex);
-    String blobName = gsUtilUri.substring(blobNameStartIndex + 1);
+    String generationGroup = m.group(3);
 
-    return BlobId.of(bucketName, blobName);
+    return BlobId.of(m.group(1), m.group(2), generationGroup == null ? null : Long.parseLong(generationGroup));
```

Comment:
Can we keep each of these argument resolutions on its own line like it was before?

i.e `String bucketName = m.group(1);`

This ensures if we ever get an exception from one of these lines we know which parameter it failed on.